### PR TITLE
testing: Add support for C source files in runner.do_run_in_out_file_test

### PR DESCRIPTION
**Pull request by sbc100**
Honor the source file extension, which means that quite a few C++ sources had
to be renamed because they were called .c.

Also, increase usage of do_run_in_out_file_test

### CODE REVIEW

File: tests/test_core.py [hunk 1]
```diff
@@ -1578,7 +1576,7 @@ def test_emscripten_run_script_string_utf8(self):
 
   def test_emscripten_get_now(self):
     self.banned_js_engines = [V8_ENGINE] # timer limitations in v8 shell
-    # needs to flush stdio streams
+# needs to flush stdio streams
```
Comment (jgravelle-google) [inline tests/test_core.py]: Indentation

File: system/include/libc/features.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #define _FEATURES_H
 
 #if defined(_ALL_SOURCE) && !defined(_GNU_SOURCE)
+#error foo
```
Comment (jgravelle-google) [inline system/include/libc/features.h]: Temporary test error?

File: tests/test_core.py (hunk 1 above)
Comment (sbc100) [reply tests/test_core.py]: Good catch!

**Comment by jgravelle-google** [comment]
I've ran into this while trying to run an individual test file manually and wondering why the behavior was different, thanks for taking the time to canonicalize it

**Pull request merged**